### PR TITLE
Github Actions testing

### PR DESCRIPTION
**Pull request by garth-wells**
- Adds Github Actions testing
- Build a single testing container using Github Actions that include the matrix of PETSc/SLEPc builds (real, complex) x (32-bit indices, 64-bit indices), including `petsc4py` and `slepc4py`. Use `PETSC_ARCH` to select the build
- Fixes for the clang/llvm 10
- A bunch of clang-tidy static analyser and style fixes

### CODE REVIEW

File: .github/workflows/ccpp.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      - name: Configure C++
+        run: cmake -G Ninja -DCMAKE_BUILD_TYPE=Developer -B build -S cpp/
+      - name: Build and install C++ library
+        run: |
+          cmake --build build
+          cmake --install build
+
+      - name: Build and run C++ unit tests (serial and MPI)
+        run: |
+          cmake --build ./build -t unittests
```
Comment (francesco-ballarin) [inline .github/workflows/ccpp.yml]: This line (and the similar one in demos) would change in #897. Let me know if you are willing to merge this before #897, and I will rebase accordingly.

File: cpp/dolfinx/io/pugixml.cpp [hunk 2]
```diff
@@ -11,24 +11,24 @@
  * Copyright (C) 2003, by Kristen Wegner (kristen@tima.net)
  */
 
-#ifndef SOURCE_PUGIXML_CPP
-#define SOURCE_PUGIXML_CPP
+#ifndef _HOME_GARTH_CODE_FENICS_DOLFINX_CPP_DOLFINX_IO_PUGIXML_CPP
```
Comment (francesco-ballarin) [inline cpp/dolfinx/io/pugixml.cpp]: This feels like a sed typo?

Comment (garth-wells) [reply cpp/dolfinx/io/pugixml.cpp]: That is a bit weird . . .

File: .github/workflows/ccpp.yml (hunk 1 above)
Comment (garth-wells) [reply .github/workflows/ccpp.yml]: I can merge this first.

Comment (francesco-ballarin) [reply .github/workflows/ccpp.yml]: OK, I will update #897 as soon as this is merged.

**Pull request merged**